### PR TITLE
Add shallow merge ability to InstanceConfigTrait::config()

### PR DESCRIPTION
**Pull request by ADmad**

### CODE REVIEW

File: src/Core/InstanceConfigTrait.php [hunk 1]
```diff
@@ -139,7 +141,11 @@ protected function _configWrite($key, $value, $merge = false)
             } else {
                 $update = [$key => $value];
             }
-            $this->_config = Hash::merge($this->_config, Hash::expand($update));
+            if ($merge === 'shallow') {
+                $this->_config = array_merge($this->_config, Hash::expand($update));
```
Comment (markstory) [inline src/Core/InstanceConfigTrait.php]: What is instead of modes there were separate methods?

File: src/Core/InstanceConfigTrait.php [hunk 2]
```diff
@@ -68,11 +68,12 @@
...
      * @param string|array|null $key The key to get/set, or a complete array of configs.
      * @param mixed|null $value The value to set.
-     * @param bool $merge Whether to merge or overwrite existing config, defaults to true.
+     * @param bool|string $merge 'deep' to merge recursively, 'shallow' for simple merge,
+     *   false to overwrite, defaults to 'deep'.
      * @return mixed Config value being read, or the object itself on write operations.
      * @throws \Cake\Core\Exception\Exception When trying to set a key that is invalid.
      */
-    public function config($key = null, $value = null, $merge = true)
+    public function config($key = null, $value = null, $merge = 'deep')
```
Comment (markstory) [inline src/Core/InstanceConfigTrait.php]: Constants would be better.

Comment (ADmad) [inline src/Core/InstanceConfigTrait.php]: Tried that. Sadly php doesn't allow constant in traits.

File: src/Core/InstanceConfigTrait.php (hunk 1 above)
Comment (ADmad) [inline src/Core/InstanceConfigTrait.php]: Don't really think it's necessary. Then we would have a method for no merge / overwriting too.

File: src/Core/InstanceConfigTrait.php (hunk 2 above)
Comment (markstory) [inline src/Core/InstanceConfigTrait.php]: That sucks :sob:

Comment (ndm2) [inline src/Core/InstanceConfigTrait.php]: Traits suck :wink: You could use an interface as a "workaround" though

```php
interface ConfigurableInterface
{
	const MERGE_DEEP = 'deep';
	
	// ...
	
	public function config($key = null, $value = null, $merge = self::MERGE_DEEP);
}

class Foo implements ConfigurableInterface
{
	use InstanceConfigTrait;
	
	public function bar() {
		echo Foo::MERGE_DEEP;
		echo PHP_EOL;
		echo self::MERGE_DEEP;
		echo PHP_EOL;
		echo static::MERGE_DEEP;
	}
}

$foo = new Foo();
$foo->bar();
```

Comment (ADmad) [inline src/Core/InstanceConfigTrait.php]: @ndm2 Yeah I found out the same workaround [here](http://stackoverflow.com/questions/24357985/php-traits-defining-generic-constants#answer-24377563). Just not too sure if we should do that..

Comment (markstory) [inline src/Core/InstanceConfigTrait.php]: I dunno if an interface is necessary. A separate method removes the need for both the constant and the interface.

Comment (dereuromark) [inline src/Core/InstanceConfigTrait.php]: I would rather support the string options than having multiple separate methods. Or what would be the naming scheme be for those methods?

Comment (markstory) [inline src/Core/InstanceConfigTrait.php]: I dunno modal options are generally a sign of a method that does too many things. We have a bunch already, bit I think it is good to think about alternatives before we add more.

Some ideas for the name are configShallow() configOverwrite() configReplace().

**Comment by markstory** [comment]
Related to #5245

File: src/Core/InstanceConfigTrait.php (hunk 2 above)
Comment (ADmad) [inline src/Core/InstanceConfigTrait.php]: @markstory What about existing config() method?

Comment (markstory) [inline src/Core/InstanceConfigTrait.php]: I find modes based on the number of arguments less troublesome than switching modes based on the value. For example combined get/set methods are less bothersome than something like App::build() in 2.x.

Comment (ADmad) [inline src/Core/InstanceConfigTrait.php]: Yes I understand your reasoning for having separate methods. My question is should the current signature of `config()` be kept same or the `$merge` argument should be removed from it in favor of separate methods.

Comment (markstory) [inline src/Core/InstanceConfigTrait.php]: I would leave the config() method as it stands for now, changing it now will cause API breaks.

**Comment by ADmad** [comment]
I reverted back the argument change to config() and added new method configShallow().

**Comment by markstory** [comment]
Looks good to me.

**Comment by dereuromark** [comment]
I wonder if this should be the default in 3.1+ or 4.0. And configDeep() then the other method name.

**Pull request merged**